### PR TITLE
BUG: MultiIndex.union dropping duplicates from result

### PR DESCRIPTION
**Pull request by phofl**
- [x] xref #38745
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

This more or less sits on top of #36299. The current base-_union implementation works only for sorted indexes correctly if indexes containing duplicates. Hence I've only added tests for sorted indexes,

### CODE REVIEW

**Review by mroeschke — APPROVED** [review]
LGTM. Any performance implications?

**Comment by phofl** [comment]
Thx, code itself is good, but this depends on #40862, after this is merged we can remove the sort_values in the tests. Will run asvs afterwards to check for performance implications

**Comment by jreback** [comment]
wait for #40862 right?

**Comment by phofl** [comment]
Yes

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.3.0.rst
Comment (phofl) [reply doc/source/whatsnew/v1.3.0.rst]: Done, will adjust here after the other is merged and ping then

**Review by jreback — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]
LGTM pending merge conflicts

**Comment by phofl** [comment]
Unfortunately this looks like a pretty high performance penalty

```
       before           after         ratio
     [84d9c5ed]       [f77120cc]
     <multiindex_union~1^2>       <multiindex_union>
+        77.4±2ms          159±4ms     2.05  multiindex_object.SetOperations.time_operation('non_monotonic', 'int', 'union')
+      33.4±0.5ms         57.9±1ms     1.73  multiindex_object.GetLoc.time_large_get_loc_warm
+         102±1ms          152±2ms     1.49  multiindex_object.SetOperations.time_operation('non_monotonic', 'string', 'union')
+      99.4±0.9ms          144±1ms     1.45  multiindex_object.SetOperations.time_operation('monotonic', 'string', 'union')
+         520±4ms         640±20ms     1.23  multiindex_object.SetOperations.time_operation('non_monotonic', 'datetime', 'union')

```

**Review by mroeschke — COMMENTED** [review]
IMO I think this is okay given that it fixes a bug. cc @jreback if this is okay despite the performance penalty.

**Comment by jreback** [comment]
where is the perf diff coming from now?

**Comment by phofl** [comment]
The Index union for duplicates is not that fast

**Comment by jreback** [comment]
ok this is fine if this slowdown only occurs when we have duplicates (which i think is the case)

**Comment by phofl** [comment]
I'll check the benchmarks

**Comment by phofl** [comment]
Unfortunately the slowdown is for also for non monotonic indexes. We could do it with only dispatching to Index.union in case of duplicates or nans, but this is makes it more complicated unfortunately.

```
       before           after         ratio
     [0158382a]       [692272cb]
                      <multiindex_union>
+         685±5ms         887±20ms     1.30  multiindex_object.SetOperations.time_operation('non_monotonic', 'datetime', 'intersection')
```

**Comment by simonjayhawkins** [comment]
@phofl can you merge master

**Comment by phofl** [comment]
Done

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**
